src/character.cpp

```diff
@@ -8951,25 +8955,58 @@ const pathfinding_settings &Character::get_pathfinding_settings() const
-                                   hammering_item.tname() );
-                int smashskill = get_arm_str() + hammering_item.damage_melee( damage_type::BASH );
-                get_map().bash( loc.position(), smashskill );
-            }
-            add_msg_if_player( _( "You crush up and gather %s" ), loc.get_item()->display_name() );
-            return true;
+    if( loc.has_parent() && loc.parent_item()->contained_where( *loc )->get_pocket_data()->rigid ) {
+        bool enough_quality = false;
+        if( has_quality( qual_DRILL ) ) {
+            enough_quality = true;
+        } else if( has_quality( qual_HAMMER ) && has_quality( qual_SCREW ) ) {
+            std::list<item_location> screw_tools;
+            std::vector<item_location> hammer_tools;
+            for( item_location &tool : all_items_loc() ) {
+                if( tool->has_quality( qual_HAMMER ) ) {
+                    hammer_tools.emplace_back( tool );
+                }
+                if( loc->has_quality( qual_SCREW ) ) {
+                    screw_tools.emplace_back( loc );
+                }
+            }
+            if( screw_tools.size() >= 1 && hammer_tools.size() >= 1 ) {
+                for( item_location &hammer : hammer_tools ) {
+                    screw_tools.remove( hammer );
+                }
+                if( screw_tools.size() >= 1 ) {
+                    enough_quality = true;
+                }
+            }
+            if( enough_quality ) {
+                add_msg_if_player( _( "You pulverize and gather %s" ), loc.get_item()->display_name() );
+                return true;
+            }
+        } else {
+            add_msg_if_player(
+                _( "You need both hammering tools and screwdrivers or a drill to crush up frozen liquids in rigid container!" ) );
         }
     } else {
-        popup( _( "You need a hammering tool to crush up frozen liquids!" ) );
+        if( has_quality( qual_HAMMER ) ) {
+            item hammering_item = item_with_best_of_quality( qual_HAMMER, true );
+            //~ %1$s: item to be crushed, %2$s: hammer name
+            if( query_yn( _( "Do you want to crush up %1$s with your %2$s?\n"
+                             "<color_red>Be wary of fragile items nearby!</color>" ),
+                          loc.get_item()->display_name(), hammering_item.tname() ) ) {
+
+                //Risk smashing tile with hammering tool, risk is lower with higher dex, damage lower with lower strength
+                if( one_in( 1 + dex_cur / 4 ) ) {
+                    add_msg_if_player( colorize( _( "You swing your %s wildly!" ), c_red ),
+                                       hammering_item.tname() );
+                    int smashskill = get_arm_str() + hammering_item.damage_melee( damage_type::BASH );
+                    get_map().bash( loc.position(), smashskill );
+                }
+                add_msg_if_player( _( "You crush up and gather %s" ), loc.get_item()->display_name() );
```

Comment:
```suggestion
                add_msg_if_player( _( "You crush up and gather %s." ), loc.get_item()->display_name() );
```

Reply:
It seems some of `character.cpp` typo were inherited by ancestor code.
I will commit these together later after fix them locally.
Thank you.